content/asm_3.md

```diff
@@ -11,243 +14,604 @@ global _start
+        ;; Put the value of the first parameter of the function from the edi register on the stack with the location rbp - 20 bytes.
+        mov     DWORD PTR [rbp-20], edi
+        ;; Put 2 to on the stack with the location rbp - 4 bytes.
+        mov     DWORD PTR [rbp-4], 2
+        ;; Put the value of the first function parameter to the eax register.
+        mov     eax, DWORD PTR [rbp-20]
+        ;; Multiple the value of the eax register to 2 and store the result in the eax register.
+        imul    eax, DWORD PTR [rbp-4]
+        ;; Restore base pointer.
+        pop     rbp
+        ;; Exit from the __dobule function.
+        ret
+```
 
-As i wroute about we have 16 general-purpose registers, and there are two interesting registers - RSP and RBP. RBP is the base pointer register. It points to the base of the current stack frame. RSP is the stack pointer, which points to the top of current stack frame.
+After the first two lines of the `__double` function, the stack frame for this function is set and looks like this:
 
-Commands
+![asm-3-stack-fram-of__double-1](./assets/asm-3-stack-of__double-1.svg)
 
-We have two commands for work with stack:
+The third instruction of the `__double` function places its first parameter to the stack with an offset of `-20`. Next, the value `2`, representing the local variable two, is also stored on the stack with an offset of `-4`. At this point, the stack frame of our function looks like this:
 
-* `push argument` - increments stack pointer (RSP) and stores argument in location pointed by stack pointer
-* `pop argument` - copied data to argument from location pointed by stack pointer
+![asm-3-stack-fram-of__double-2](./assets/asm-3-stack-of__double-2.svg)
 
-Let's look on one simple example:
+Finally, we put the value from the stack at offset `-20` (the value of the function's parameter) into the `eax` register and multiply it by `2`, which is located on the stack at offset `-4`. The result of the multiplication is then stored in the `eax` register. This simple example shows how the stack is used to access both parameters and local variables of a function.
 
-```assembly
-global _start
+## Stack operations
 
-section .text
+We already have seen two assembly instructions that affects the current state of the stack:
 
-_start:
-		mov rax, 1
-		mov rdx, 2
-		push rax
-		push rdx
+- `push` - pushes the operand into the stack.
+- `pop` - pops the top value from the stack.
 
-		mov rax, [rsp + 8]
+x86_64 processors provide additional instruction that brings affect on the stack. Besides those instruction we also have seen familiar to us:
 
-		;;
-		;; Do something
-		;;
+- `call`
+- `ret`
+
+The first one instruction calls the given procedure. It affects stack by saving the return address on the stack before call. The second instruction is an "exit" from the given procedure. It affects the stack by removing the return address from the stack and transferring the execution flow to it. 
```

Comment:
```suggestion

```

I would explain these instructions right in the list as in the comment above. If you like this idea, please remove this paragraph.

Reply:
Will remove